_studio/mfx_lib/scheduler/src/mfx_scheduler_core_ischeduler.cpp

```diff
@@ -362,63 +361,50 @@ mfxStatus mfxSchedulerCore::WaitForTaskCompletion(const void *pOwner)
         WakeUpThreads();
     }
 
+    std::list<mfxTaskHandle> tasks;
 
-    do
     {
-        // reset waiting handle
-        waitHandle.handle = 0;
-
-        // enter protected section.
-        // make searching in a separate code block,
-        // to avoid dead-locks with other threads.
-        {
-            std::lock_guard<std::mutex> guard(m_guard);
-            int priority;
+        mfxTaskHandle waitHandle;
+        std::lock_guard<std::mutex> guard(m_guard);
 
-            priority = MFX_PRIORITY_HIGH;
-            do
+        ForEachTask(
+            [&](MFX_SCHEDULER_TASK* task)
             {
-                int type;
-
-                for (type = MFX_TYPE_HARDWARE; type <= MFX_TYPE_SOFTWARE; type += 1)
+                //make a list of all 'active' tasks of given owner
+                if ((task->param.task.pOwner == pOwner) && (MFX_WRN_IN_EXECUTION == task->opRes))
                 {
-                    const MFX_SCHEDULER_TASK *pTask = m_pTasks[priority][type];
-
-                    // look through tasks list. Find the specified object.
-                    while (pTask)
-                    {
-                        // the task with equal ID has been found.
-                        if ((pTask->param.task.pOwner == pOwner) &&
-                            (MFX_WRN_IN_EXECUTION == pTask->opRes))
-                        {
-                            // create wait handle to make waiting more precise
-                            waitHandle.taskID = pTask->taskID;
-                            waitHandle.jobID = pTask->jobID;
-                            break;
-                        }
-
-                        // get the next task
-                        pTask = pTask->pNext;
-                    }
+                    waitHandle.taskID = task->taskID;
+                    waitHandle.jobID = task->jobID;
+
+                    tasks.emplace_back(waitHandle);
                 }
 
-            } while ((0 == waitHandle.handle) && (--priority >= MFX_PRIORITY_LOW));
+                return false;
+            }
+        );
+    }
+
+    auto handle = tasks.begin();
+    while (!tasks.empty())
+    {
+        if (tasks.end() == handle)
+            handle = tasks.begin();
 
-            // leave the protected section
+        // wait for a while
+        // for a while and infinite are different things
+        if (MFX_WRN_IN_EXECUTION == Synchronize(*handle, MFX_TIME_TO_WAIT)) // Still working
```

Comment:
Is it possible to stuck here forever? With some hanged task returning ```MFX_WRN_IN_EXECUTION``` over and over

Reply:
I think yes. but it was original logic. I don't work to change behaviour here